src/logic/programmable-function-logic.ts

```diff
@@ -63,39 +63,47 @@ export class ProgrammableFunctionLogic extends WatchableFunctionLogic {
     this.answerByIndex = {};
   }
 
-  private async modifyAnswer(result: EVMResult): Promise<void> {
+  private async modifyAnswer(result: EVMResult, call: ContractCall): Promise<void> {
     const answer = this.answerByIndex[this.getCallCount() - 1] || this.defaultAnswer;
 
     if (answer) {
       result.gasUsed = new BN(0);
-      result.execResult.returnValue = answer.encodedValue;
       result.execResult.gasUsed = new BN(0);
-      result.execResult.exceptionError = answer.shouldRevert ? new VmError('lopt revert' as any) : undefined;
+      if (answer.shouldRevert) {
+        result.execResult.exceptionError = new VmError('lopt revert' as any);
+        result.execResult.returnValue = this.encodeRevertReason(answer.value);
+      } else {
+        result.execResult.returnValue = await this.encodeValue(answer.value, call);
+      }
     }
   }
 
-  private encodeValue(value?: ProgrammedReturnValue): Buffer {
+  private async encodeValue(value: ProgrammedReturnValue, call: ContractCall): Promise<Buffer> {
     if (value === undefined) return EMPTY_ANSWER;
 
-    let encodedReturnValue: string = '0x';
+    let toEncode = typeof value === 'function' ? await value(call.args) : value;
+
+    let encoded: string = '0x';
```

Comment:
Nit: I think `encodedReturnValue` was a clearer variable name here

Reply:
I agree! 😄 Changed it in the next commit